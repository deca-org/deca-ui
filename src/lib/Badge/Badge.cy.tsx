import { Test } from '../Utils';
import _cyp from '../../../cypress';
import React from 'react';
import Badge from './Badge';
import { standardColors } from '../Theme';

describe('components/Badge', () => {
  const selector = '[data-testid="test.badge"]';
  it('pill', () => {
    cy.mount(
      <Badge pill data-testid="test.badge">
        Label
      </Badge>
    );
    cy.get(selector).should(
      'have.css',
      'border-radius',
      Test.borderRadius('pill')
    );
  });
  describe('sizes', () => {
    it('sm', () => {
      cy.mount(
        <Badge size="sm" data-testid="test.badge">
          Label
        </Badge>
      );
      cy.get(selector).should('have.css', 'fontSize', Test.fontSize('caption'));
      cy.get(selector).should('have.css', 'paddingTop', Test.space('1'));
      cy.get(selector).should('have.css', 'paddingBottom', Test.space('1'));
      cy.get(selector).should('have.css', 'paddingLeft', Test.space('2'));
      cy.get(selector).should('have.css', 'paddingRight', Test.space('2'));
    });
    it('md', () => {
      cy.mount(
        <Badge size="md" data-testid="test.badge">
          Label
        </Badge>
      );
      cy.get(selector).should('have.css', 'fontSize', Test.fontSize('bodySm'));
      cy.get(selector).should('have.css', 'paddingTop', Test.space('1'));
      cy.get(selector).should('have.css', 'paddingBottom', Test.space('1'));
      cy.get(selector).should('have.css', 'paddingLeft', Test.space('3 - 1'));
      cy.get(selector).should('have.css', 'paddingRight', Test.space('3 - 1'));
    });
    it('lg', () => {
      cy.mount(
        <Badge size="lg" data-testid="test.badge">
          Label
        </Badge>
      );
      cy.get(selector).should('have.css', 'fontSize', Test.fontSize('body'));
      cy.get(selector).should('have.css', 'paddingTop', Test.space('1'));
      cy.get(selector).should('have.css', 'paddingBottom', Test.space('1'));
      cy.get(selector).should('have.css', 'paddingLeft', Test.space('3'));
      cy.get(selector).should('have.css', 'paddingRight', Test.space('3'));
    });
  });
  describe('color', () => {
    standardColors.map((color) => {
      it(color, () => {
        cy.mount(
          <Badge color={color} data-testid="test.badge">
            Label
          </Badge>
        );
        cy.get(selector).should(
          'have.css',
          'background-color',
          Test.color(color)
        );
        cy.get(selector).should(
          'have.css',
          'color',
          Test.color(`${color}-readable`)
        );
      });
    });
  });
});
