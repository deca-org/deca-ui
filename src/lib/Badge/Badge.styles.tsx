import { styled } from '@lib/Theme/stitches.config';

export const StyledBadge = styled('span', {
  transition: '$default',
  fontFamily: '$normal',
  fontWeight: '$bold',
  variants: {
    pill: {
      true: {
        br: '$pill',
      },
      false: {
        br: '$sm',
      },
    },
    size: {
      sm: {
        fontSize: '$caption',
        py: '$1',
        px: '$2',
      },
      md: {
        fontSize: '$bodySm',
        py: '$1',
        paddingRight: 'calc($3 - $1)',
        paddingLeft: 'calc($3 - $1)',
      },
      lg: {
        fontSize: '$body',
        py: '$1',
        px: '$3',
      },
    },
    color: {
      primary: {
        bg: '$primary',
        color: '$primary-readable',
      },
      secondary: {
        bg: '$secondary',
        color: '$secondary-readable',
      },
      success: {
        bg: '$success',
        color: '$success-readable',
      },
      warning: {
        bg: '$warning',
        color: '$warning-readable',
      },
      error: {
        bg: '$error',
        color: '$error-readable',
      },
    },
  },
});
