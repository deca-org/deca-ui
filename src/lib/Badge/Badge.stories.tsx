import { DecaUIProvider } from '@lib/Theme';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import React from 'react';

import Badge from './Badge';

export default {
  title: 'Badge',
  component: Badge,
} as ComponentMeta<typeof Badge>;

const Template: ComponentStory<typeof Badge> = (args) => <Badge {...args} />;

export const Default = Template.bind({});
Default.args = {
  children: 'Hello',
  size: 'md',
  color: 'primary',
  pill: false,
  css: {},
  className: '',
};

export const WithTheme = Template.bind({});

WithTheme.args = { ...Default.args };
WithTheme.decorators = [
  (Story) => (
    <DecaUIProvider
      theme={{
        colors: {
          primary: '$green600',
        },
        radii: {
          sm: '10px',
        },
      }}
    >
      <Story />
    </DecaUIProvider>
  ),
];
