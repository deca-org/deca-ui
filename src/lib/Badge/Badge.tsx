import { CSS, StandardColors } from '@lib/Theme/stitches.config';
import {
  __DEV__,
  PolymorphicRef,
  PolymorphicComponentPropWithRef,
} from '@lib/Utils';
import clsx from 'clsx';
import React from 'react';

import { StyledBadge } from './Badge.styles';

/**
 * Badges are used to highlight an item's status for quick recognition.
 */
interface Props {
  /**
   * The content of the component.
   */
  children?: React.ReactNode | undefined;
  /**
   * ClassName applied to the component.
   * @default ''
   */
  className?: string;
  /**
   * Color to use.
   * @default primary
   */
  color?: StandardColors;
  /**
   * Size of the component.
   * @default md
   */
  size?: 'sm' | 'md' | 'lg';
  /**
   * Override default CSS style.
   */
  css?: CSS;
  /**
   * Have component be pill shaped
   */
  pill?: boolean;
}

export type BadgeProps<T extends React.ElementType> =
  PolymorphicComponentPropWithRef<T, Props>;

export type BadgeComponent = (<C extends React.ElementType = 'div'>(
  props: BadgeProps<C>
) => React.ReactElement | null) & { displayName?: string };

const Badge: BadgeComponent = React.forwardRef(
  <T extends React.ElementType = 'div'>(
    {
      children,
      className,
      color = 'primary',
      size = 'md',
      as,
      css,
      pill = false,
      ...props
    }: BadgeProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const preClass = 'decaBadge';

    return (
      <StyledBadge
        ref={ref}
        size={size}
        as={as}
        css={css}
        color={color}
        className={clsx(className, `${preClass}-root`)}
        pill={pill}
        {...props}
      >
        {children}
      </StyledBadge>
    );
  }
);

if (__DEV__) {
  Badge.displayName = 'DecaUI.Badge';
}

export default Badge;
