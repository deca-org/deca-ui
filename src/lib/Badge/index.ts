import Badge from './Badge';
export * from './Badge';
export { Badge };
export default Badge;
