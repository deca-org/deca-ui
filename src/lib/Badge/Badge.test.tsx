import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import React from 'react';

import Badge from './Badge';

describe('components/Badge', () => {
  it('renders content', () => {
    const { getByText } = render(<Badge>Label</Badge>);
    expect(getByText('Label')).toBeInTheDocument();
  });
  it('matches snapshot', () => {
    const { asFragment } = render(<Badge />);
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders all colors', () => {
    const { asFragment } = render(
      <>
        <Badge color="primary" />
        <Badge color="secondary" />
        <Badge color="success" />
        <Badge color="warning" />
        <Badge color="error" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders all sizes', () => {
    const { asFragment } = render(
      <>
        <Badge size="sm" />
        <Badge size="md" />
        <Badge size="lg" />
      </>
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it('renders pill option', () => {
    const { asFragment } = render(<Badge pill />);
    expect(asFragment()).toMatchSnapshot();
  });
});
