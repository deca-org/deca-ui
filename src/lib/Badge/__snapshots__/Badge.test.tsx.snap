// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Badge matches snapshot 1`] = `
<DocumentFragment>
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-kVJBgm-size-md c-eCAFWP-kKaAKY-color-primary decaBadge-root"
  />
</DocumentFragment>
`;

exports[`components/Badge renders all colors 1`] = `
<DocumentFragment>
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-kVJBgm-size-md c-eCAFWP-kKaAKY-color-primary decaBadge-root"
  />
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-kVJBgm-size-md c-eCAFWP-fDgeqI-color-secondary decaBadge-root"
  />
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-kVJBgm-size-md c-eCAFWP-cnyuIA-color-success decaBadge-root"
  />
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-kVJBgm-size-md c-eCAFWP-fLkJis-color-warning decaBadge-root"
  />
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-kVJBgm-size-md c-eCAFWP-efKikA-color-error decaBadge-root"
  />
</DocumentFragment>
`;

exports[`components/Badge renders all sizes 1`] = `
<DocumentFragment>
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-ioXeqN-size-sm c-eCAFWP-kKaAKY-color-primary decaBadge-root"
  />
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-kVJBgm-size-md c-eCAFWP-kKaAKY-color-primary decaBadge-root"
  />
  <span
    class="c-eCAFWP c-eCAFWP-fzQAVN-pill-false c-eCAFWP-jBFCiy-size-lg c-eCAFWP-kKaAKY-color-primary decaBadge-root"
  />
</DocumentFragment>
`;

exports[`components/Badge renders pill option 1`] = `
<DocumentFragment>
  <span
    class="c-eCAFWP c-eCAFWP-euJLgm-pill-true c-eCAFWP-kVJBgm-size-md c-eCAFWP-kKaAKY-color-primary decaBadge-root"
  />
</DocumentFragment>
`;
